src/api/replace.cc

```diff
@@ -0,0 +1,346 @@
+            newRanges->PushBack(methodArguments.transaction->GetRange(range->start + offset, \
+                range->end + offset, \
+                range->inputInfo)); \
+        } \
+    } \
+    subjectIt++; \
+}
+
+void TaintReplaceStringByStringMethod(const FunctionCallbackInfo<Value>& args) {
+    if (!utils::ValidateMethodArguments(args, 6, "Wrong number of arguments")) {
+        return;
+    }
+    auto replaceResult = args[1];
+    if (!replaceResult->IsString()) {
+        args.GetReturnValue().Set(replaceResult);
+        return;
+    }
+    try {
+        auto methodArguments = ReadStringMatcherMethodArguments(args);
+        if (methodArguments.transaction == nullptr) {
+            args.GetReturnValue().Set(methodArguments.result);
+            return;
+        }
+
+        auto subjectRanges = GetTaintedRangesFromLocal(methodArguments.transaction, methodArguments.self);
+        auto replacerRanges = GetTaintedRangesFromLocal(methodArguments.transaction, methodArguments.replacer);
+        if (subjectRanges == nullptr && replacerRanges == nullptr) {
+            args.GetReturnValue().Set(replaceResult);
+            return;
+        }
+
+        INIT_STRING_MATCHER_PROPERTIES
+
+        std::vector<Range*>::iterator subjectIt;
+        std::vector<Range*>::iterator subjectItEnd;
+        if (subjectRanges) {
+            subjectIt = subjectRanges->begin();
+            subjectItEnd = subjectRanges->end();
+            ADD_BEFORE_REPLACEMENT_RANGES
+        }
+
+        ADD_REPLACEMENT_RANGES
+
+        if (subjectRanges) {
+            ADD_AFTER_REPLACEMENT_RANGES
+        }
+
+        if (newRanges->Size() > 0) {
+            auto isolate = args.GetIsolate();
+            auto resultString = replaceResult->ToString(isolate->GetCurrentContext()).ToLocalChecked();
+            auto resultLength = resultString->Length();
+            if (resultLength == 1) {
+                replaceResult = tainted::NewExternalString(isolate, replaceResult);
+            }
+            auto key = utils::GetLocalStringPointer(replaceResult);
+            methodArguments.transaction->AddTainted(key, newRanges, replaceResult);
+        }
+    } catch (const std::bad_alloc& err) {
+    } catch (const container::QueuedPoolBadAlloc& err) {
+    } catch (const container::PoolBadAlloc& err) {
+    }
+    args.GetReturnValue().Set(replaceResult);
+}
+#undef INIT_STRING_MATCHER_PROPERTIES
+#undef ADD_BEFORE_REPLACEMENT_RANGES
+#undef ADD_REPLACEMENT_RANGES
+#undef ADD_AFTER_REPLACEMENT_RANGES
+
+#define INIT_REGEX_REPLACE_PROPERTIES \
+auto isolate = args.GetIsolate(); \
+auto context = isolate->GetCurrentContext(); \
+auto jsReplacements = v8::Array::Cast(*replacements); \
+auto jsReplacementsLength = jsReplacements->Length(); \
+auto replacerLength = v8::String::Cast(*(methodArguments.replacer))->Length(); \
+std::vector<Range*>::iterator subjectIt; \
+std::vector<Range*>::iterator subjectItEnd; \
+
+#define ADD_RANGES_BEFORE_REPLACEMENT \
```

Comment:
Same as in String flavor. Maybe it's worth exploring having the same function for both flavors by just having one structure and add a property indicating the type of the structure. Like:
`struct MatcherArguments {
    enum_matcher_t type;  // MATCHER_TYPE_STRING or MATCHER_TYPE_REGEX
    Transaction* transaction;
    v8::Local<Value> result;
    v8::Local<Value> self;
    v8::Local<Value> matcher;
    v8::Local<Value> replacer;
    v8::Local<Value> replacement;
};`
This way the function can accept both types and switch-case the algorithm used based on the type.